File: 42-trapping-rain-water/trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        long sum = 0;
        int len = height.size();

        int arrLeft[len];
        int arrRight[len];

        int maxL = 0;
        int maxR = 0;

        // potential water
        for(int i=0 ; i < len ;i++) {
            arrLeft[i] = maxL;
            
            if (maxL < height[i]) {
                maxL = height[i];
            }
        }

        // potential water from right
        for(int i=len-1 ; i >= 0 ;i--) {
            arrRight[i] = maxR;
            
            if (maxR < height[i]) {
                maxR = height[i];
            }
        }

        // calculate
        for(int i=0 ;i < len ; i++) {
            sum += max(min(arrLeft[i], arrRight[i]) - height[i], 0);
        }

        return sum;
    }
};

File: 42-trapping-rain-water/trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int[] leftH = new int[height.length];
        int[] rightH = new int[height.length];
        
        int maxHeight = height[0];
        for(int i = 0 ; i< height.length ; i++) {
            maxHeight = Math.max(maxHeight, height[i]);
            leftH[i] = maxHeight;
        }

        maxHeight = height[height.length-1];
        for(int i = height.length-1 ; i >= 0 ; i--) {
            maxHeight = Math.max(maxHeight, height[i]);
            rightH[i] = maxHeight;
        }

        int ans = 0;
        for(int i = 0 ; i< height.length ; i++) {
            ans +=  Math.max(Math.min(rightH[i], leftH[i]) - height[i], 0);
        }

        return ans;
    }
}